### PR TITLE
[ENHANCEMENT] Use an existing expectation_store.store_backend_id to initialize an in-memory DataContext

### PR DESCRIPTION
**Pull request by anthonyburdi**
Changes proposed in this pull request:
- add _construct_data_context_id to choose the id of the currently-configured expectations store, if available. If not, choose the id stored in DataContextConfig. Keep existing functionality intact e.g. setting data_context_id via env var.
- manually init `expectations_store.store_backend_id` if there is a data_context_id in DataContextConfig and no expectation store backend is configured already
- tests

@jcampbell for design review
@talagluck for code review

### CODE REVIEW

**Comment by anthonyburdi** [comment]
Working on fixing failing CI tests, will re-request reviews when passing.

File: great_expectations/data_context/data_context.py
Comment (talagluck) [inline great_expectations/data_context/data_context.py]: Is there a reason you use `except Exception:` instead of `except:` here?

File: great_expectations/data_context/data_context.py
Comment (jcampbell) [inline great_expectations/data_context/data_context.py:468]: There is a small red flag for me in grabbing an id through a substitution-aware property during handling of an exception. Inside the exception handling, I think we need to do as little as possible that could go wrong, especially since we're catching any exception (e.g. even keyboard interrupt, etc)

File: great_expectations/data_context/data_context.py
Comment (jcampbell) [inline great_expectations/data_context/data_context.py]: I think it would be preferrable to do this only once, and then set the data_context_id to persist throughout the life of the instance

File: great_expectations/data_context/data_context.py
Comment (jcampbell) [inline great_expectations/data_context/data_context.py]: See above comment -> would prefer to only construct the first time it is accessed.

File: great_expectations/data_context/data_context.py
Comment (talagluck) [inline great_expectations/data_context/data_context.py]: Could we add a slightly more detailed explanation? The if statement conditions don't seem to match the comment - I understand that they are the same thing, but it could be confusing.

File: great_expectations/data_context/data_context.py
Comment (talagluck) [inline great_expectations/data_context/data_context.py]: awesome!

**Review by talagluck — APPROVED** [review]
Just a suggestion and a question, but I don't want to hold up approval.

File: great_expectations/data_context/data_context.py
Comment (jcampbell) [inline great_expectations/data_context/data_context.py]: Can you clarify what "manually_initialize_store_backend_id" does? I don't follow yet.

Comment (anthonyburdi) [reply great_expectations/data_context/data_context.py]: Sure! From the docstring in [StoreBackend.store_backend_id()](https://github.com/great-expectations/great_expectations/pull/2046/files#diff-3e42f11c8884077bdeff67d7316330c4f63a792505d8019629dde7dd39af2a07R43)
```
        If a valid UUID store_backend_id is passed in param manually_initialize_store_backend_id
        and there is not already an existing store_backend_id then the store_backend_id
        from param manually_initialize_store_backend_id is used to create it.
```
This is so we can manually set the store_backend_id, e.g. when the expectations_store does not yet exist, but the data_context_id is set in the project_config like in [this test](https://github.com/great-expectations/great_expectations/pull/2046/files#diff-c3b5b880409a33e00a2a5f1bb2655ba5cf9e14ec6ec92ced6237a0907f4634faR241) or the env var GE_DATA_CONTEXT_ID is set like in [this test](https://github.com/great-expectations/great_expectations/pull/2046/files#diff-c3b5b880409a33e00a2a5f1bb2655ba5cf9e14ec6ec92ced6237a0907f4634faR295).

I'm also very open to renaming it to something more descriptive!

File: great_expectations/data_context/data_context.py
Comment (anthonyburdi) [reply great_expectations/data_context/data_context.py]: Nope, good catch!

File: great_expectations/data_context/data_context.py
Comment (anthonyburdi) [reply great_expectations/data_context/data_context.py]: ~There is a possibility this is not necessary, looking into it.~ See following comment

Comment (anthonyburdi) [reply great_expectations/data_context/data_context.py]: My mistake - this is necessary. I forgot that I'm using the `manually_initialize_store_backend_id` parameter here to set the `expectations_store.store_backend_id` to the `project_config.anonymous_usage_statistics.data_context_id` if the expectations_store doesn't yet exist.

File: great_expectations/data_context/data_context.py
Comment (anthonyburdi) [reply great_expectations/data_context/data_context.py]: This try/except is now not necessary due to swapping the order of initializing the stores and the data_context_id. If necessary, we could have changed `Exception` to `AttributeError`

File: great_expectations/data_context/data_context.py
Comment (anthonyburdi) [reply great_expectations/data_context/data_context.py]: I made an edit in https://github.com/great-expectations/great_expectations/pull/2046/commits/868a0b3183ccc14b4663383835782d2896281ca8 - let me know if it makes sense or happy to edit / add!

File: great_expectations/data_context/data_context.py
Comment (anthonyburdi) [reply great_expectations/data_context/data_context.py:468]: Yes - I definitely should have changed to `AttributeError` at a minimum. The good news is that this try/except is not needed after swapping the order in `__init__` to instantiate the stores first before constructing the data_context_id

File: great_expectations/data_context/data_context.py
Comment (anthonyburdi) [reply great_expectations/data_context/data_context.py]: Done - let me know if you see any further issues!

File: great_expectations/data_context/data_context.py
Comment (anthonyburdi) [reply great_expectations/data_context/data_context.py]: Done - let me know if you see any further issues!

**Review by jcampbell — APPROVED** [review]
Thanks for the explanations!

**Pull request merged**
